client/src/components/ElementEditor/ElementList.js

```diff
@@ -11,6 +11,101 @@ import { getDragIndicatorIndex } from 'lib/dragHelpers';
 import { getElementTypeConfig } from 'state/editor/elementConfig';
 
 class ElementList extends Component {
+  constructor(props) {
+    super(props);
+    this.resetState = this.resetState.bind(this);
+    this.handleBeforeSubmitForm = this.handleBeforeSubmitForm.bind(this);
+    this.handleAfterSubmitResponse = this.handleAfterSubmitResponse.bind(this);
+    this.state = {
+      // saveAllElements will be set to true in entwine.js in the 'onbeforesubmitform' "hook"
+      // which is triggered by LeftAndMain submitForm()
+      saveAllElements: false,
+      // increment is also set in entwine.js in the 'onbeforesubmitform' "hook"
+      increment: 0,
+      hasUnsavedChangesBlockIDs: {},
+      validBlockIDs: {},
+    };
+    // Update the sharedObject so that setState() can be called from entwine.js
+    this.props.sharedObject.setState = this.setState.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // Number of blocks just changed after graphql response
+    if (this.props.blocks !== prevProps.blocks) {
+      this.resetState(prevState, false);
+      return;
+    }
+    // Saving all elements and state has just updated because of a formSchema response from inline save
+    if (this.state.saveAllElements) {
+      const unsavedChangesBlockIDs = this.props.blocks
+        .map(block => parseInt(block.id, 10))
+        .filter(blockID => this.state.hasUnsavedChangesBlockIDs[blockID]);
+      const allValidated = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID] !== null);
+      if (allValidated) {
+        const allValid = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID]);
+        // entwineResolve is bound in entwine.js
+        this.props.sharedObject.entwineResolve(allValid);
+        this.resetState(prevState, allValid);
+        this.setState({ saveAllElements: false });
+      }
+    }
+  }
+
+  resetState(prevState, resetHasUnsavedChangesBlockIDs) {
+    if (this.props.blocks === null) {
+      return;
+    }
+    // validBlockIDs (though not hasUnsavedChangesBlockIDs) uses a tri-state
+    // - null: not saved
+    // - true: saved, valid
+    // - false: attempted save, invalid
+    const hasUnsavedChangesBlockIDs = {};
+    const validBlockIDs = {};
+    this.props.blocks.forEach(block => {
+      const blockID = parseInt(block.id, 10);
+      if (resetHasUnsavedChangesBlockIDs) {
+        hasUnsavedChangesBlockIDs[blockID] = false;
+      } else if (prevState.hasUnsavedChangesBlockIDs.hasOwnProperty(blockID)) {
+        hasUnsavedChangesBlockIDs[blockID] = prevState.hasUnsavedChangesBlockIDs[blockID];
+      } else {
+        hasUnsavedChangesBlockIDs[blockID] = false;
+      }
+      validBlockIDs[blockID] = null;
```

Comment:
So we're saying all blocks are "not saved" in valid block IDs, but there's a separate `hasUnsavedChangesBlockIDs` as well? I'm really confused here. Do we have two states both tracking the saved/unsaved state of blocks?
And why does "validBlockIDs" track saved status? Shouldn't it just track include valid IDs and exclude everything else like the name implies?

This might just be a naming thing but this is really confusing me.

Reply:
hasUnsavedChangesBlockIDs is the blocks dirty state - true/false

validBlockIDs is the blocks validation state - true/false/null(i.e. unknown - haven't sent and/or received response from server)

Reply:
Ahh okay. That makes sense after you explained it.... I don't know what we can do about the variable names to make it clearer... can you please add a comment somewhere to that effect so we have that context immediately clear when looking at the code?

Reply:
@emteknetnz Please respond to this

Reply:
Have added inline comments